README


This archive contains the distribution MooseX-AlwaysCoerce,
version 0.21:

  Automatically enable coercions for Moose attributes

This software is copyright (c) 2009 by Rafael Kitover <rkitover@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for MooseX-AlwaysCoerce

0.21      2013-11-30 18:13:01Z
    - re-release to fix compile test

0.20      2013-09-17 01:32:52Z (Karen Etheridge)
    - fixed bad test plan

0.19      2013-09-17 01:29:18Z (Karen Etheridge)
    - warnings tests bypassed during installation
    - repository has moved to the GitHub Moose organization

0.18      2012-10-13 16:57:07 PDT-0700 (Karen Etheridge)
    - added missing prerequisite declaration
    - added lots more metadata; kwalitee should be improved

0.17      2012-06-01 17:52:39 UTC+0000
    - conversion to Dist::Zilla

0.16 2011-03-16
    - bump minimum required Package::Stash and namespace::autoclean versions
      in dependencies, to reduce invalid cpantesters reports. (The toolchain
      already reports on conflicting Package::Stash, but not everything
      is catching that.)

0.15 2011-03-14
    - bump minimum required MooseX::ClassAttribute version in dependencies, to
      ensure compatibility with the upcoming Moose 2.0 release.

0.14 2011-03-08
    - add support within roles, using the new applied_attribute feature
      available in Moose 1.99/2.0 (doy)

0.13 2011-03-07
    - fixed class attribute and parameterized role TODO tests to work again
      with the latest MooseX::Role::Parameterized (Karen Etheridge)
    - converted tests to Test::Fatal (Karen Etheridge)

0.12  2010-10-08 12:14:54 PDT
    - ...and again.

0.11  2010-10-08 12:05:42 PDT
    - fix mangled distribution tarball (Karen Etheridge)

0.10  2010-10-08 10:03:53 PDT
    - avoid giving class attributes to the metaclass, which was triggering an
      edge case bug in metaclass compat (Jesse Luehrs)

0.09  2010-08-05 03:42:07
    - support attributes without isa, correctly

0.08  2010-08-05 03:14:37
    - make a test safer

0.07  2010-08-05 03:10:12
    - support class_has with no isa (Karen Etheridge)

0.06  2010-07-28 00:58:50
    - Use modern Moose APIs, to avoid warnings with Moose 1.09 (Karen Etheridge)

0.05  2010-07-13 00:05:10
    - support types without coercion (Schwern)

0.04  2010-03-06 05:28:01
    - remove t/00-load.t (was failing)

0.03  2009-06-16 17:50:12
    - make coerce => 0 work for class_has

0.02  2009-06-16 17:23:29
    - rewritten using metaroles, the Moose way

0.01  2009-06-15 22:18:34
    - releasing...


lib/MooseX/AlwaysCoerce.pm
package MooseX::AlwaysCoerce;
{
  $MooseX::AlwaysCoerce::VERSION = '0.21';
}
# git description: v0.20-7-gd590377

BEGIN {
  $MooseX::AlwaysCoerce::AUTHORITY = 'cpan:RKITOVER';
}
# ABSTRACT: Automatically enable coercions for Moose attributes

use strict;
use warnings;

use namespace::autoclean 0.12;
use Moose ();
use MooseX::ClassAttribute 0.24 ();
use Moose::Exporter;
use Moose::Util::MetaRole;
use Carp;

Moose::Exporter->setup_import_methods;


{
    package MooseX::AlwaysCoerce::Role::Meta::Attribute;
{
  $MooseX::AlwaysCoerce::Role::Meta::Attribute::VERSION = '0.21';
}
# git description: v0.20-7-gd590377

BEGIN {
  $MooseX::AlwaysCoerce::Role::Meta::Attribute::AUTHORITY = 'cpan:RKITOVER';
}
    use namespace::autoclean;
    use Moose::Role;

    around should_coerce => sub {
        my $orig = shift;
        my $self = shift;

        my $current_val = $self->$orig(@_);

        return $current_val if defined $current_val;

        return 1 if $self->type_constraint && $self->type_constraint->has_coercion;
        return 0;
    };

    package MooseX::AlwaysCoerce::Role::Meta::Class;
{
  $MooseX::AlwaysCoerce::Role::Meta::Class::VERSION = '0.21';
}
# git description: v0.20-7-gd590377

BEGIN {
  $MooseX::AlwaysCoerce::Role::Meta::Class::AUTHORITY = 'cpan:RKITOVER';
}
    use namespace::autoclean;
    use Moose::Role;
    use Moose::Util::TypeConstraints;

    around add_class_attribute => sub {
        my $next = shift;
        my $self = shift;
        my ($what, %opts) = @_;

        if (exists $opts{isa}) {
            my $type = Moose::Util::TypeConstraints::find_or_parse_type_constraint($opts{isa});
            $opts{coerce} = 1 if not exists $opts{coerce} and $type->has_coercion;
        }

        $self->$next($what, %opts);
    };
}

my (undef, undef, $init_meta) = Moose::Exporter->build_import_methods(

    install => [ qw(import unimport) ],

    class_metaroles => {
        attribute   => ['MooseX::AlwaysCoerce::Role::Meta::Attribute'],
        class       => ['MooseX::AlwaysCoerce::Role::Meta::Class'],
    },

    role_metaroles => {
        (Moose->VERSION >= 1.9900
            ? (applied_attribute => ['MooseX::AlwaysCoerce::Role::Meta::Attribute'])
            : ()),
        role                => ['MooseX::AlwaysCoerce::Role::Meta::Class'],
    }
);

sub init_meta {
    my ($class, %options) = @_;
    my $for_class = $options{for_class};

    MooseX::ClassAttribute->import({ into => $for_class });

    # call generated method to do the rest of the work.
    goto $init_meta;
}

1;
# vim:et sts=4 sw=4 tw=0:

__END__

=pod

=encoding UTF-8

=for :stopwords Rafael Kitover <rkitover@cpan.org> Jesse Luehrs Karen Etheridge Michael G.
Schwern coercions AnnoCPAN Rolsky

=head1 NAME

MooseX::AlwaysCoerce - Automatically enable coercions for Moose attributes

=head1 VERSION

version 0.21

=head1 SYNOPSIS

    package MyClass;

    use Moose;
    use MooseX::AlwaysCoerce;
    use MyTypeLib 'SomeType';

    has foo => (is => 'rw', isa => SomeType); # coerce => 1 automatically added

    # same, MooseX::ClassAttribute is automatically applied
    class_has bar => (is => 'rw', isa => SomeType);

=head1 DESCRIPTION

Have you ever spent an hour or more trying to figure out "Hey, why did my
coercion not run?" only to find out that you forgot C<< coerce => 1 >> ?

Just load this module in your L<Moose> class and C<< coerce => 1 >> will be
enabled for every attribute and class attribute automatically.

Use C<< coerce => 0 >> to disable a coercion explicitly.

=for Pod::Coverage init_meta

=head1 BUGS

Please report any bugs or feature requests to C<bug-moosex-alwayscoerce at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=MooseX-AlwaysCoerce>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find more information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=MooseX-AlwaysCoerce>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/MooseX-AlwaysCoerce>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/MooseX-AlwaysCoerce>

=item * Search CPAN

L<http://search.cpan.org/dist/MooseX-AlwaysCoerce/>

=back

=head1 ACKNOWLEDGEMENTS

My own stupidity, for inspiring me to write this module.

Dave Rolsky, for telling me how to do it the L<Moose> way.

=head1 AUTHOR

Rafael Kitover <rkitover@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2009 by Rafael Kitover <rkitover@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 CONTRIBUTORS

=over 4

=item *

Jesse Luehrs <doy@tozt.net>

=item *

Karen Etheridge <ether@cpan.org>

=item *

Michael G. Schwern <schwern@pobox.com>

=back

=cut


Makefile.PL
# This Makefile.PL for MooseX-AlwaysCoerce was generated by
# Dist::Zilla::Plugin::MakeMaker::Fallback 0.004.
# Don't edit it but the dist.ini used to construct it.

use strict;
use warnings;

BEGIN {
my %configure_requires = (
    'ExtUtils::MakeMaker' => '6.30',
    'Module::Build::Tiny' => '0.030',
);

my @missing = grep {
    ! eval "require $_; $_->VERSION($configure_requires{$_}); 1"
} keys %configure_requires;

if (not @missing)
{
    print "Congratulations, your toolchain understands 'configure_requires'!\n\n";
}
else
{
    warn <<'EOW';
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

If you're installing manually, please retrain your fingers to run Build.PL
when present instead.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

EOW

    sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));
}
}

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Automatically enable coercions for Moose attributes",
  "AUTHOR" => "Rafael Kitover <rkitover\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build::Tiny" => "0.030"
  },
  "DISTNAME" => "MooseX-AlwaysCoerce",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "MooseX::AlwaysCoerce",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Moose" => 0,
    "Moose::Exporter" => 0,
    "Moose::Role" => 0,
    "Moose::Util::MetaRole" => 0,
    "Moose::Util::TypeConstraints" => 0,
    "MooseX::ClassAttribute" => "0.24",
    "Package::Stash" => "0.26",
    "namespace::autoclean" => "0.12",
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec::Functions" => 0,
    "List::Util" => 0,
    "Test::Fatal" => 0,
    "Test::More" => 0,
    "Test::Requires" => 0,
    "if" => 0
  },
  "VERSION" => "0.21",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



